doc/neps/ufunc-overrides.rst

```diff
@@ -106,163 +116,193 @@ Returning ``NotImplemented`` to user should not happen. Moreover::
   immediately.
 
-If none of the input arguments has a ``__numpy_ufunc__`` method, the
+If none of the input arguments has an ``__array_ufunc__`` method, the
 execution falls back on the default ufunc behaviour.
 
+Subclass hierarchies
+--------------------
+
+Hierarchies of such containers (say, a masked quantity), are most easily
+constructed if methods consistently use :func:`super` to pass through
+the class hierarchy [7]_.  To support this, :class:`ndarray` has its own
+``__array_ufunc__`` method (which is equivalent to ``getattr(ufunc,
+method)(*inputs, **kwargs)``, i.e., if any of the (adjusted) inputs
+still defines ``__array_ufunc__`` that will be called in turn). This
+should be particularly useful for container-like subclasses of
+:class:`ndarray`, which add an attribute like a unit or mask to a
+regular :class:`ndarray`. Such classes can do possible adjustment of the
+arguments relevant to their own class, pass on to another class in the
+hierarchy using :func:`super` until the Ufunc is actually done, and then
+do possible adjustments of the outputs.
+
+Turning Ufuncs off
+------------------
+
+For some classes, Ufuncs make no sense, and, like for other special
+methods [8]_, one can indicate Ufuncs are not available by setting
+``__array_ufunc__`` to :obj:`None`.  Inside a Ufunc, this is
+equivalent to unconditionally return :obj:`NotImplemented`, and thus
+will lead to a :exc:`TypeError` (unless another operand implements
+``__array_ufunc__`` and knows how to deal with the class).
+
+.. [7] https://rhettinger.wordpress.com/2011/05/26/super-considered-super/
+
+.. [8] https://docs.python.org/3/reference/datamodel.html#specialnames
 
 In combination with Python's binary operations
 ----------------------------------------------
 
-The ``__numpy_ufunc__`` mechanism is fully independent of Python's
+The ``__array_ufunc__`` mechanism is fully independent of Python's
 standard operator override mechanism, and the two do not interact
 directly.
 
-They however have indirect interactions, because NumPy's ``ndarray``
-type implements its binary operations via Ufuncs. Effectively, we have::
-
-    class ndarray(object):
+They have indirect interactions, however, because NumPy's
+:class:`ndarray` type implements its binary operations via Ufuncs.  For
+most numerical classes, the easiest way to override binary operations is
+thus to define ``__array_ufunc__`` and override the corresponding
+Ufunc. The class can then, like :class:`ndarray` itself, define the
+binary operators in terms of Ufuncs. Here, one has to take some care.
+E.g., the simplest implementation would be::
+
+    class ArrayLike(object):
+        def __array_ufunc__(self, ufunc, method, *inputs, **kwargs):
+            ...
+            return result
         ...
         def __mul__(self, other):
-            return np.multiply(self, other)
+            return self.__array_ufunc__(np.multiply, '__call__', self, other)
```

Comment:
Also, `return self.__array_ufunc__(...)` vs. `if getattr(other, '__array_ufunc__', False) is None: return NotImplemented; else return np.multiply(self, other)` correspond to different opt-out approaches in the previous discussion. The former is the laissez-faire "option 3" that allows defining `__array_ufunc__` while doing opt-out, whereas latter does not allow opt-out and `__array_ufunc__` simultaneously ("option 2", IIRC).

What to recommend for subclasses in the NEP should follow what `ndarray` itself does --- IIUC,  this PR decided to do "option 2" since there's discussion of the None special value.

The other options should be discussed in a separate section "Other proposals considered", and only give one way to do it as an example...